hyperspy/misc/io/fei_stream_readers.py

```diff
@@ -219,77 +55,147 @@ def stream_to_sparse_COO_array(
+        last_index = last_frame*ydim*xdim
+    else:
+        last_frame = frms
+    filt = (pixel_indexes>=first_index) & (pixel_indexes<last_index)
+    pixel_indexes = pixel_indexes[filt]
+    channels = channels[filt]
+    # if the frames are to be summed the pixel indexes are updated
+    final_shape = (last_frame-first_frame, ydim, xdim,
+                       chnx//rebin_energy)
+    coords = (pixel_indexes//(xdim*ydim)-first_frame,
+                  pixel_indexes%(xdim*ydim)//xdim,
+                  pixel_indexes%(xdim*ydim)%xdim,
+                  channels)
+    counts = np.ones(channels.shape[0], dtype=stream.dtype)
+    return coords, counts, final_shape
 
 
 @jit_ifnumba()
-def _fill_array_with_stream(spectrum_image, stream, first_frame,
-                            last_frame, rebin_energy=1):
-    navigation_index = 0
-    frame_number = 0
-    shape = spectrum_image.shape
-    for count_channel in np.nditer(stream):
-        # when we reach the end of the frame, reset the navigation index to 0
-        if navigation_index == (shape[1] * shape[2]):
-            navigation_index = 0
-            frame_number += 1
-            # break the for loop when we reach the last frame we want to read
-            if frame_number == last_frame:
-                break
-        # if different of ‘65535’, add a count to the corresponding channel
-        if count_channel != 65535:
-            if first_frame <= frame_number:
-                spectrum_image[frame_number - first_frame,
-                               navigation_index // shape[2],
-                               navigation_index % shape[2],
-                               count_channel // rebin_energy] += 1
-        else:
-            navigation_index += 1
+def _stream_to_sparse_summed(stream, shape, rebin_energy=1,
+                                       first_frame=0, last_frame=None):
+    """
+    Vectorized implementation of stream to sparse conversion. Returns
+    coordinates, data, and shape to be turned to a COO matrix
+
+    Parameters
+    ----------
+    stream: np.array
+        Velox emd stream data
+    shape: (number of frames, y dimension, x dimension, number of channels)
+        number of frames can be none
+    rebin_energy: int, optional
+        factor to rebin energy axis, must be divisor of number of channels
+    first_frame: int, optional
+        first frame to consider
+    last_frame: int, optional
+        last frame to consider
+    summed: bool, optional
+        Add up data from all frames to one spectrum image
+
+    Returns
+    -------
+    coords : (ydim, xdim, channel) array
+        coordinates where counts are to be added. shape depends on summed.
+    counts : 1D array
+        just an array of ones representing counts
+    final_shape: 3 tuple
+        (ydim, xdim, channel)
+    """
+    frms, ydim, xdim, chnx = shape
+    # the indexes where counts are registered
+    count_channel_regs = np.argwhere(stream != 65535)[:, 0]
```

Comment:
What shape is `stream`? Is it 1D?

What does `np.flatnonzero(stream != 65535)` instead of the `argwhere`?

Reply:
Tried it out, cool function that I didn't know about. No significant speed-up unfortunately neither with a sparse nor full matrix.